crates/autopilot/src/domain/settlement/trade.rs

```diff
@@ -39,102 +43,491 @@ impl Trade {
         }
     }
 
+    /// CIP38 score defined as surplus + protocol fee
+    ///
+    /// Denominated in NATIVE token
+    #[allow(dead_code)]
+    fn score(
+        &self,
+        prices: &auction::Prices,
+        policies: &[fee::Policy],
+    ) -> Result<eth::Ether, Error> {
+        tracing::debug!("Scoring trade {:?}", self);
+        Ok(self.native_surplus(prices)? + self.native_protocol_fee(prices, policies)?)
+    }
+
+    /// A general surplus function.
+    ///
+    /// Can return different types of surplus based on the input parameters.
+    ///
+    /// Denominated in SURPLUS token
+    fn surplus_over(
+        &self,
+        prices: &ClearingPrices,
+        price_limits: PriceLimits,
+    ) -> Result<eth::Asset, error::Math> {
+        match self.side {
+            order::Side::Buy => {
+                // scale limit sell to support partially fillable orders
+                let limit_sell = price_limits
+                    .sell
+                    .0
+                    .checked_mul(self.executed.into())
+                    .ok_or(error::Math::Overflow)?
+                    .checked_div(price_limits.buy.0)
+                    .ok_or(error::Math::DivisionByZero)?;
+                let sold = self
+                    .executed
+                    .0
+                    .checked_mul(prices.buy)
+                    .ok_or(error::Math::Overflow)?
+                    .checked_div(prices.sell)
+                    .ok_or(error::Math::DivisionByZero)?;
+                limit_sell.checked_sub(sold).ok_or(error::Math::Negative)
+            }
+            order::Side::Sell => {
+                // scale limit buy to support partially fillable orders
+                let limit_buy = self
+                    .executed
+                    .0
+                    .checked_mul(price_limits.buy.0)
+                    .ok_or(error::Math::Overflow)?
+                    .checked_div(price_limits.sell.0)
+                    .ok_or(error::Math::DivisionByZero)?;
+                let bought = self
+                    .executed
+                    .0
+                    .checked_mul(prices.sell)
+                    .ok_or(error::Math::Overflow)?
+                    .checked_ceil_div(&prices.buy)
+                    .ok_or(error::Math::DivisionByZero)?;
```

Comment:
Being able to convert amounts from one side of the trade to another side of the trade (e.g. some amount in buy tokens to that amount in sell tokens) seems to be quite a common operation here (I see it at least 3 times). Maybe we can expose this in a more easy to use way that can be shared?

Reply:
Tried but not happy with what I ended up with, particularly because types are not identical in these code repetitions.